### PR TITLE
Replace use of Linux only regex.h with STL <regex> 

### PR DESCRIPTION
**Pull request by hassec**
This replaces all the existing use of [`regex.h`](https://man7.org/linux/man-pages/man0/regex.h.0p.html) with the [`<regex>`](https://en.cppreference.com/w/cpp/header/regex) header available in the standard library since C++11.   

All tests pass, thus the common use cases seem to be covered. 
The syntax is set to POSIX so should match the old behavior, though we could consider changing it to e.g. ECMAScript which maybe is by now more common?

The biggest benefit of this change is that Windows should now no longer be a 2nd class citizen when it comes to the `--grep` option.

### CODE REVIEW

File: cmake/generateConfigFile.cmake [hunk 1]
```diff
@@ -39,7 +39,6 @@ int main() {
 check_include_file_cxx( "unistd.h"      EXV_HAVE_UNISTD_H )
 check_include_file_cxx( "sys/mman.h"    EXV_HAVE_SYS_MMAN_H )
 if ( NOT MINGW AND NOT MSYS AND NOT MSVC )
-check_include_file_cxx( "regex.h"       EXV_HAVE_REGEX_H )
 endif()
```
Comment (kevinbackhouse) [inline cmake/generateConfigFile.cmake:42]: Can lines 41 and 42 be removed too? It looks like an empty if now.

**Comment by clanmills** [comment]
Gentlemen.  If you're doing maintenance on the regex code, can you reconsider the design.  I accepted the contribution because I hoped to encourage and recruit another engineer.

I didn't like that what was done.   It's not just the name which is:

```
const exv_grep_keys_t& greps
```
It should be something like `class Exiv2::Grep;`  However, it's the lack of abstraction.  The class should encapsulate all its working.  For example in src/version.cpp the code executes `regexec()` in-line.   `regexec()` should be executed in the Exiv2::Grep class.  Ditto for g->bIgnoreCase_.

The code is messy because some C++98 platforms (msvc) do not provide regular expression handling.  I believe that's always available in C++11.

Time for a damn good cleanup and simplification. 

```cpp
static bool shouldOutput(const exv_grep_keys_t& greps,const char* key,const std::string& value)
{
    bool bPrint = greps.empty();
    for (auto g = greps.begin(); !bPrint && g != greps.end(); ++g) {
        std::string Key(key);
#if defined(EXV_HAVE_REGEX_H)
        bPrint = (0 == regexec(&(*g), key, 0, nullptr, 0) || 0 == regexec(&(*g), value.c_str(), 0, nullptr, 0));
#else
            std::string Pattern(g->pattern_);
            std::string Value(value);
            if ( g->bIgnoreCase_ ) {
                // https://notfaq.wordpress.com/2007/08/04/cc-convert-string-to-upperlower-case/
                std::transform(Pattern.begin(), Pattern.end(),Pattern.begin(), ::tolower);
                std::transform(Key.begin()    , Key.end()    ,Key.begin()    , ::tolower);
                std::transform(Value.begin()  , Value.end()  ,Value.begin()    , ::tolower);
            }
            bPrint = Key.find(Pattern) != std::string::npos || Value.find(Pattern) != std::string::npos;
#endif
    }
    return bPrint;
}
```

File: cmake/generateConfigFile.cmake (hunk 1 above)
Comment (hassec) [reply cmake/generateConfigFile.cmake]: Yes, totally! Thanks for spotting it :blush:

**Review by kevinbackhouse — DISMISSED** [review]

**Comment by hassec** [comment]
I've added one more commit that removes the type alias `exv_grep_keys_t`.  
`std::vector<std::regex>` is IMHO better as a reader directly knows what one is dealing with.   
Especially as in this case the type is plain STL stuff, so this means people will directly know what they can do with a `std::vector<std::regex>`


As far as the cleanup comment goes from @clanmills, I'd say that this PR does exactly this.  
After all this PR removes 191 lines and only adds 34.  
Regarding adding something along `class Exiv2::Grep`, I'm not sure if I'm convinced we need that additional layer of abstraction.     
I quite like how few lines of code the current solution needs, and that means it's easy to maintain :smile:     

But I'm happy to be convinced otherwise.  I might just be missing something. 
@kevinbackhouse what do you think?

**Review by kevinbackhouse — APPROVED** [review]

**Pull request merged**